Optimize parts of the entity cramming check

magneticflux- opened the pull request:
Some preliminary results:
```
All tests using 365 chickens on a ladder in an empty world after a fresh restart and a 3600-tick warp.

Pre-patch:
	Trial 1:
		Total:		12.565ms
		Chicken:	11.77ms
	Trial 2:
		Total:		12.235ms
		Chicken:	11.45ms
	Trial 3:
		Total:		12.432ms
		Chicken:	11.63ms
	Average:
		Total:		12.411ms
		Chicken:	11.62ms

a4f0ef38:
	Trial 1:
		Total:		9.613ms
		Chicken:	8.56ms
	Trial 2:
		Total:		9.299ms
		Chicken:	8.35ms
	Trial 3:
		Total:		9.312ms
		Chicken:	8.38ms
	Average:
		Total:		9.408ms
		Chicken:	8.43ms

f97444b6:
	Trial 1:
		Total:		7.781ms
		Chicken:	6.72ms
	Trial 2:
		Total:		7.605ms
		Chicken:	6.62ms
	Trial 3:
		Total:		7.478ms
		Chicken:	6.57ms
	Average:
		Total:		7.621ms
		Chicken:	6.64ms

ba1b3e48:
	Trial 1:
		Total:		8.251ms
		Chicken:	7.25ms
	Trial 2:
		Total:		8.153ms
		Chicken:	7.22ms
	Trial 3:
		Total:		8.158ms
		Chicken:	7.24ms
	Average:
		Total:		8.187ms
		Chicken:	7.24ms

525609dd:
	Trial 1:
		Total:		7.652ms
		Chicken:	6.70ms
	Trial 2:
		Total:		7.565ms
		Chicken:	6.69ms
	Trial 3:
		Total:		7.505ms
		Chicken:	6.61ms
	Average:
		Total:		7.574ms
		Chicken:	6.67ms

5589b3d5:
	Trial 1:
		Total:		7.139ms
		Chicken:	6.11ms
	Trial 2:
		Total:		6.961ms
		Chicken:	6.04ms
	Trial 3:
		Total:		6.876ms
		Chicken:	5.96ms
	Average:
		Total:		6.992ms
		Chicken:	6.04ms

Average Improvement:
	a4f0ef38:
		Total:		-3.003ms	-24.2%
		Chicken:	-3.19ms		-27.5%
	f97444b6:
		Total:		-4.790ms	-38.6%
		Chicken:	-4.98ms		-42.9%
	ba1b3e48:
		Total:		-4.224ms	-34.0%
		Chicken:	-4.38ms		-37.7%
	525609dd:
		Total:		-4.837ms	-39.0%
		Chicken:	-4.95ms		-42.6%
	5589b3d5:
		Total:		-5.419ms	-43.7%
		Chicken:	-5.58ms		-48.0%

```

magneticflux- committed:
new: entity.cached_climbing_check: Don't update the climbing status of every entity for every colliding entity each frame

Just uses the previous value, which was calculated earlier the same tick so should be correct. Helps with the n^2 complexity of collisions a little bit.

magneticflux- committed:
Avoid method call in fast path by using object equality

Technically someone _could_ reflection-hack a second `Optional` object with a `null` value, but that isn't supported.

magneticflux- committed:
fix: Explicitly cache climbing value and update it each tick

magneticflux- committed:
change: Cache all calls to `isClimbing` per-tick

magneticflux- committed:
new: cache calls to `getHealth`

Forces and update once per tick, also updates on `setHealth`.

magneticflux- commented:
I think this is as optimized as it gets without rewriting the collision system to act per-world (using some kind of disjoint-set or octree) instead of per-entity.

2No2Name commented:
Is this PR still relevant, as similar things (the same thing?) was merged to carpet mod?

magneticflux- commented:
Yes, I believe this still provides a statistically significant improvement in TPS while not obviously breaking vanilla parity like the Carpet mod does.

2No2Name commented:
I believe this PR breaks vanilla parity in cases where entities collide with climbable blocks that are placed or removed. Therefore this cannot be merged

2No2Name commented:
After a discussion on discord we agreed on trying to add better cache invalidation to this PR to resolve the problems.
1) Invalidate when the entity moves
2) Invalidate when the chunk changes

magneticflux- committed:
fix: Invalidate climbing cache if position is set or if the chunk section has been modified

This resolves https://github.com/CaffeineMC/lithium-fabric/pull/260#issuecomment-903217788

magneticflux- commented on src/main/java/me/jellysquid/mods/lithium/mixin/entity/cached_climbing_check/MixinLivingEntity.java:
```diff
@@ -0,0 +1,56 @@
+    @Inject(method = "isClimbing", cancellable = true, at = @At("HEAD"))
+    private void useCacheIsClimbing(CallbackInfoReturnable<Boolean> cir) {
+        int blockY = this.getBlockY();
+        if (!world.isOutOfHeightLimit(blockY)) {
+            Chunk chunk = (Chunk) ((CollisionView) world).getChunkAsView(Pos.ChunkCoord.fromBlockCoord(this.getBlockX()), Pos.ChunkCoord.fromBlockCoord(this.getBlockZ()));
+            if (chunk != null) {
+                SectionModCounter section = (SectionModCounter) chunk.getSectionArray()[Pos.SectionYIndex.fromBlockCoord(world, blockY)];
```
I don't know if there's a better way to do this yet. I based it on `LandPathNodeMakerMixin` here: https://github.com/CaffeineMC/lithium-fabric/blob/fde14ea8281ca2f5233ad6a5b395a74ca3315cf6/src/main/java/me/jellysquid/mods/lithium/mixin/ai/pathing/LandPathNodeMakerMixin.java#L70-L79

2No2Name replied:
section definitely needs a null check here

magneticflux- committed:
fix: Handle empty sections

2No2Name commented:
I guess this can be profiled yet another time now

magneticflux- commented:
Didn't have the original setup, but I redid some of it from scratch:
```
All tests using 512 chickens on a ladder in an empty world after a fresh restart and a 3600-tick warp.

Pre-patch:
	Trial 1:
		Total:		25.331ms
		Chicken:	22.84ms
	Trial 2:
		Total:		25.698ms
		Chicken:	23.14ms
	Trial 3:
		Total:		25.380ms
		Chicken:	22.85ms
	Average:
		Total:		25.470ms
		Chicken:	22.94ms

a69f8082:
	Trial 1:
		Total:		17.647ms
		Chicken:	14.42ms
	Trial 2:
		Total:		17.430ms
		Chicken:	14.34ms
	Trial 3:
		Total:		17.030ms
		Chicken:	14.21ms
	Average:
		Total:		17.369ms
		Chicken:	14.32ms

Average Improvement:
	a69f8082:
		Total:		-8.101ms	-31.8%
		Chicken:	-8.62ms		-37.5%
```

2No2Name commented:
Can confirm, blinking redstone lines went from 41mspt to 31mspt.

magneticflux- committed:
change: only invalidate the climbing cache if moved to a whole new block

2No2Name commented:
I will profile the isClimbing caching. The health caching just seems to duplicate the information from the datatracker to avoid the datatracker. If that turns out to be a good idea, I would like to find out whether that can be applied to other tracked data as well (e.g. item stack of item entities).

2No2Name commented:
I just noticed that the ladder system still has some flaws:

- LivingEntity.setPosition is not the only place where the position is modified
- Trapdoors on top of ladders are like ladders only if there is a ladder below. The ladder below can be in a different chunk section, and therefore the cache should be invalidated when the ladder is broken, but isn't.

2No2Name commented:
![2021-08-23_19 20 58](https://user-images.githubusercontent.com/50278648/130490297-fc18029c-20d4-41b7-a7ce-d933f174ac08.png) shows a setup that demonstrates the problem. Climbing up, then breaking the ladders below, will lead to the player still climbing the trapdoors, even though they are no longer climbable. Same test without the patch will cause the player to drop when the ladder is broken

2No2Name commented:
I will fix these problems and profile it then

2No2Name commented:
Pushed my changes here. https://github.com/2No2Name/lithium-fabric/tree/feature/entity-cramming The performance is less than 10% better in my testing

2No2Name commented:
Closing due to the newer better PR
